src/storage.zig

```diff
@@ -7,400 +7,555 @@ const log = std.log.scoped(.storage);
+                // TODO What if we receive a temporary EIO error because of a faulty cable?
+                error.InputOutput => @panic("latent sector error: no spare sectors to reallocate"),
+                // TODO: It seems like it might be possible for some filesystems to return ETIMEDOUT
+                // here. Consider handling this without panicking.
+                else => {
+                    log.err(
+                        "impossible write: offset={} buffer.len={} error={s}",
+                        .{ write.offset, write.buffer.len, @errorName(err) },
+                    );
+                    @panic("impossible write");
+                },
+            };
+
+            if (bytes_written == 0) {
+                // This should never happen if the kernel and filesystem are well behaved.
+                // However, block devices are known to exhibit this behavior in the wild.
+                // TODO: Consider retrying with a timeout if this panic proves problematic, and be
+                // careful to avoid logging in a busy loop. Perhaps a better approach might be to
+                // return wrote = null here and let the protocol retry at a higher layer where
+                // there is more context available to decide on how important this is or whether
+                // to cancel.
+                @panic("write operation returned 0 bytes written");
+            }
+
+            write.offset += bytes_written;
+            write.buffer = write.buffer[bytes_written..];
+
+            if (write.buffer.len == 0) {
+                write.callback(write);
+                return;
+            }
+
+            self.start_write(write);
         }
-    }
 
-    pub fn read_sectors(
-        self: *Storage,
-        callback: *const fn (read: *Storage.Read) void,
-        read: *Storage.Read,
-        buffer: []u8,
-        zone: vsr.Zone,
-        offset_in_zone: u64,
-    ) void {
-        zone.verify_iop(buffer, offset_in_zone);
-        assert(zone != .grid_padding);
-
-        const offset_in_storage = zone.offset(offset_in_zone);
-        read.* = .{
-            .completion = undefined,
-            .callback = callback,
-            .buffer = buffer,
-            .offset = offset_in_storage,
-            .target_max = buffer.len,
-        };
+        /// Ensures that the read or write is within bounds and intends to read or write some bytes.
+        fn assert_bounds(self: *Self, buffer: []const u8, offset: u64) void {
+            _ = self;
+            _ = offset;
 
-        self.start_read(read, null);
-        assert(read.target().len > 0);
-    }
+            assert(buffer.len > 0);
+        }
+    };
+}
 
-    fn start_read(self: *Storage, read: *Storage.Read, bytes_read: ?usize) void {
-        assert(read.offset % constants.sector_size == 0);
-        maybe(bytes_read == 0); // Retrying erroneous read at the same offset with smaller window.
+test "storage with mocked io layer" {
```

Comment:
So this test seemed to run quite quickly - do you think it has enough coverage as is? Or would it benefit from being a standalone fuzz (like all the other _fuzz.zig) fuzzers?

(I ask, since if it would benefit from more time, then we likely want to split it out to keep our unit tests quick)

Reply:
Very good point. I think turning it into a standalone fuzz would make the most sense. Will address everything else first and then take a go at that. :)